plotly/figure_factory/_facet_grid.py

```diff
@@ -0,0 +1,1073 @@
+    ```
+
+    Example 3: Categorical Coloring
+    ```
+    import plotly.plotly as py
+    import plotly.figure_factory as ff
+
+    import pandas as pd
+
+    mpg = pd.read_table('https://raw.githubusercontent.com/plotly/datasets/master/mpg_2017.txt')
+
+    fig = ff.create_facet_grid(
+        mtcars,
+        x='mpg',
+        y='wt',
+        facet_col='cyl',
+        color_name='cyl',
+        color_is_cat=True,
+    )
+    py.iplot(fig, filename='facet_grid_mpg_default_colors')
+    ```
+
+    Example 4: Sequential Coloring
+    ```
+    import plotly.plotly as py
+    import plotly.figure_factory as ff
+
+    import pandas as pd
+
+    tips = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/tips.csv')
+
+    fig = ff.create_facet_grid(
+        tips,
+        x='total_bill',
+        y='tip',
+        facet_row='sex',
+        facet_col='smoker',
+        color_name='size',
+        colormap='Viridis',
+    )
+    py.iplot(fig, filename='facet_grid_tips_sequential_colors')
+    ```
+
+    Example 5: Custom labels
+    ```
+    import plotly.plotly as py
+    import plotly.figure_factory as ff
+
+    import pandas as pd
+
+    mtcars = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/mtcars.csv')
+
+    fig = ff.create_facet_grid(
+        mtcars,
+        x='wt',
+        y='mpg',
+        facet_col='cyl',
+        facet_col_labels={4: "$\\alpha$", 6: '$\\beta$', 8: '$\sqrt[y]{x}$'},
+    )
+
+    py.iplot(fig, filename='facet_grid_mtcars_custom_labels')
+    ```
+    """
+    if not pd:
+        raise exceptions.ImportError(
+            "'pandas' must be imported for this figure_factory."
+        )
+
+    if not isinstance(df, pd.DataFrame):
+        raise exceptions.PlotlyError(
+            "You must input a pandas DataFrame."
+        )
+
+    # make sure all columns are of homogenous datatype
+    utils.validate_dataframe(df)
+
+    for key in [x, y, facet_row, facet_col, color_name]:
+        if key is not None:
+            try:
+                df[key]
+            except KeyError:
+                raise exceptions.PlotlyError(
+                    "x, y, facet_row, facet_col and color_name must be keys "
+                    "in your dataframe."
+                )
+
+    # validate scales
+    if scales not in ['fixed', 'free_x', 'free_y', 'free']:
+        raise exceptions.PlotlyError(
+            "'scales' must be set to 'fixed', 'free_x', 'free_y' and 'free'."
+        )
+
+    if trace_type not in ['scatter', 'scattergl']:
+        raise exceptions.PlotlyError(
+            "'trace_type' must be 'scatter' or 'scattergl'."
+        )
+
+    # seperate kwargs for marker and else
+    if 'marker' in kwargs:
+        kwargs_marker = kwargs['marker']
+    else:
+        kwargs_marker = {}
+    for param in ['size', 'color']:
+        kwargs_marker.pop(param, None)
+    kwargs.pop('marker', None)
+
+    # make sure dataframe index starts at 0
+    df.index = range(len(df))
```

Comment:
I'm not sure we want to be doing this: this is going to be modifying the user's dataframe without them knowing.

For example:
![image](https://cloud.githubusercontent.com/assets/1280389/26708315/1e5ea8dc-4719-11e7-84a2-43d49c1e51a8.png)